src/generic_datatype.cpp

```diff
@@ -0,0 +1,88 @@
+#include "generic_datatype.h"
+
+#include <memory>
+#include <unordered_map>
+
+#include "assert.h"
+
+using namespace std;
+
+namespace smt {
+
+GenericDatatype::GenericDatatype(const DatatypeDecl & dt_declaration,
+                                 const std::string & s)
+    : dt_decl(dt_declaration), name(s)
+{
+}
+
+GenericDatatypeConstructorDecl::GenericDatatypeConstructorDecl(
+    const std::string & name)
+    : cons_name(name)
+{
+}
+void GenericDatatypeConstructorDecl::add_new_selector(
+    const shared_ptr<selectorComponents> & newSelector)
+{
+  selector_vector.push_back(*newSelector);
+}
+
+std::vector<selectorComponents>
+GenericDatatypeConstructorDecl::get_selector_vector()
+{
+  return selector_vector;
+}
+
+void GenericDatatype::add_constructor(
+    const GenericDatatypeConstructorDecl & dt_cons_decl)
+{
+  cons_decl_vector.push_back(dt_cons_decl);
+}
+
+std::string GenericDatatype::get_name() const { return name; }
+
+int GenericDatatypeConstructorDecl::get_selector_count() const
+{
+  return selector_vector.size();
+}
+
+int GenericDatatype::get_num_constructors() const
+{
+  return cons_decl_vector.size();
+}
+
+int GenericDatatype::get_num_selectors(std::string cons) const
+{
+  int num_selectors = 0;
+  for (unsigned int i = 0; i < cons_decl_vector.size(); ++i)
+  {
+    if (cons_decl_vector[i].get_name() == cons)
+    {
+      num_selectors = cons_decl_vector[i].get_selector_count();
+    }
+  }
+  return num_selectors;
+}
+
+void GenericDatatype::add_selector(
+    const GenericDatatypeConstructorDecl & dt_cons_decl,
+    const std::shared_ptr<selectorComponents> & newSelector)
+{
+  for (unsigned int i = 0; i < cons_decl_vector.size(); ++i)
+  {
+    if (cons_decl_vector[i].get_name() == dt_cons_decl.get_name())
```

Comment:
The fact that we chose to compare datatypes etc. by their names is an implementation detail, that will be hard to remember every time we want to compare them. So I suggest following the `AbsSort` and `GenericSort` pattern. `sort.h` overrides the `==` operator. `sort.h` implements this by calling the `compare` function, that can be implemented by various classes that inherit from `AbsSort`. In particular, `GenericDatatype` implements such a `compare` function. And when you call `==` between two generic datatypes, since `GenericDatatype` does not override the operator `==`, the code of `AbsSort` is called, and hence `compare` is executed. 

Note that this will require adding a `dataype.cpp` file.

Let me know if anything of this isn't clear!

Reply:
So basically, follow the model set by sort and generic sort which entails reconfiguring datatype.h and a new datatype.cpp to override the '==' function in the same way as sort.cpp, namely calling compare on the objects. And then implementing a compare function in genericdatatype (which could also just compare the names)?

Reply:
yes!

Reply:
As a bit of an update, changing datatype.h and cpp might affect the generic cvc4 implementation. I will know more concretely after a bit as some other lurking problemsthat I hope yo have addressed soon in my implementation might be causing the issues.

Reply:
I suggest that for now, just configure smt-switch without cvc4. Maybe `./configure.sh --debug --z3`. This should still test the generic solver's datatype implementation (using `z3 as a back-end instead of `cvc4`), and that way errors regarding cvc4 will disappear (for now).

If you think you can resolve this, go for it, but I wouldn't waste much time on it -- we can for sure fix that later after the generic solver works with creating datatypes sorts.